c_src/workitems.h

```diff
@@ -264,391 +220,275 @@ namespace eleveldb {
+//=======================================================================
 
-    class RangeScanTask : public WorkTask
-    {
+struct RangeScanOptions {
+
+    // Byte-level controls for batching/ack
+    
+    size_t max_unacked_bytes;
+    size_t low_bytes;
+    size_t max_batch_bytes;
+    
+    // Max number of items to return. Zero means unlimited.
+    
+    size_t limit;
+    
+    // Include the start key in streaming iteration?
+    
+    bool start_inclusive;
+    
+    // Include the end key in streaming iteration?
+    
+    bool end_inclusive;
+    
+    // Read options
+    
+    bool fill_cache;
+    bool verify_checksums;
+    
+    // Filter options
+    
+    Encoding::Type encodingType_;
+    ERL_NIF_TERM rangeFilterSpec_;
+    ErlNifEnv* env_;
+    bool useRangeFilter_;
+    
+    RangeScanOptions();
+    ~RangeScanOptions();
+    
+    //------------------------------------------------------------
+    // Sanity-check filter options
+    //------------------------------------------------------------
+    
+    void checkOptions();
+    
+};  // struct RangeScanOptions
+    
+class RangeScanTask : public WorkTask
+{
+public:
+
+    // Used to coordinate production and consumption of batches of data.
+    // Producers acknowledge each batch received. Consumers block when the
+    // unacked limit has been reached and need to be woken up by the consumer.
+    // When consumers die, the ref count is decremented and that will signal
+    // the producer to go away too.
+
+    class SyncObject : public RefObject {
     public:
-        // Used to coordinate production and consumption of batches of data.
-        // Producers acknowledge each batch received. Consumers block when the
-        // unacked limit has been reached and need to be woken up by the consumer.
-        // When consumers die, the ref count is decremented and that will signal
-        // the producer to go away too.
-        class SyncObject : public RefObject {
-        public:
-            explicit SyncObject(const RangeScanOptions & opts);
-            ~SyncObject();
-
-            // True if only one side (producer or consumer) alive.
-            inline bool SingleOwned() { return m_RefCount == 1; }
-
-            // Adds number of bytes sent to count.
-            // Will block if count exceeds max waiting for the other
-            // side to ack some and take it under the limit or for the other
-            // side to shut down.
-            void AddBytes(uint32_t n);
-
-            void AckBytes(uint32_t n);
-            bool AckBytesRet(uint32_t n);
-
-            // Should be called when the Erlang handle is garbage collected
-            // so no process is there to consume the output.
-            void MarkConsumerDead();
-
-            bool IsConsumerDead() const;
-
-        private:
-            const uint32_t max_bytes_;
-            const uint32_t low_bytes_;
-            volatile uint32_t num_bytes_;
-            bool producer_sleeping_;
```

Comment:
Indeed.  Done.

On Mon, Oct 19, 2015 at 2:12 PM, Matthew Von-Maszewski <
notifications@github.com> wrote:

> In c_src/workitems.h
> <https://github.com/basho/eleveldb/pull/162#discussion_r42426597>:
>
> > -            void AddBytes(uint32_t n);
> > -
> > -            void AckBytes(uint32_t n);
> > -            bool AckBytesRet(uint32_t n);
> > -
> > -            // Should be called when the Erlang handle is garbage collected
> > -            // so no process is there to consume the output.
> > -            void MarkConsumerDead();
> > -
> > -            bool IsConsumerDead() const;
> > -
> > -        private:
> > -            const uint32_t max_bytes_;
> > -            const uint32_t low_bytes_;
> > -            volatile uint32_t num_bytes_;
> > -            bool producer_sleeping_;
>
> Should be volatile so that compiler knows it cannot optimize value into
> register. Is used in tight loop around a condition variable to communicate
> state between threads.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/basho/eleveldb/pull/162/files#r42426597>.
>